commerce/domain/shipping/cart/calculator/classes/ShippingCartCalculatorSample.cls

```diff
@@ -39,54 +39,57 @@ public class ShippingCartCalculatorSample extends CartExtension.ShippingCartCalc
     } else {
       CartExtension.CartItemList cartItems = cart.getCartItems();
       Integer numberOfUniqueItems = cartItems.size();
-      CartExtension.CartDeliveryGroup cartDeliveryGroup = cartDeliveryGroups.get(0);
-      CartExtension.CartDeliveryGroupMethodList cartDeliveryGroupMethods = cartDeliveryGroup.getCartDeliveryGroupMethods();
-
-      // Clean up the CartDeliveryGroupMethods
-      for (Integer i = (cartDeliveryGroupMethods.size() - 1); i >= 0; i--) {
-        CartExtension.CartDeliveryGroupMethod method = cartDeliveryGroupMethods.get(i);
-        cartDeliveryGroupMethods.remove(method);
-      }
-      // To clear selected Cart Delivery Group Method
-      cartDeliveryGroup.setSelectedCartDeliveryGroupMethod(null);
-
-      // Get the Shipping Product
-      List<Product2> shippingProducts = [SELECT Id FROM Product2 WHERE ProductClass != 'VariationParent' LIMIT 1];
-      if(shippingProducts.size() == 0) {
-        CartExtension.CartValidationOutput cvo = new CartExtension.CartValidationOutput(CartExtension.CartValidationOutputTypeEnum.SHIPPING,
-                CartExtension.CartValidationOutputLevelEnum.ERROR );
-        cvo.setMessage('No Shipping Products have been defined');
-        cartValidationOutputList.add(cvo);
-      } else {
+      
+      for (Integer i = (cartDeliveryGroups.size() - 1); i >= 0; i--) {
+        CartExtension.CartDeliveryGroup cartDeliveryGroup = cartDeliveryGroups.get(i);
+        CartExtension.CartDeliveryGroupMethodList cartDeliveryGroupMethods = cartDeliveryGroup.getCartDeliveryGroupMethods();
+
+        // Clean up the CartDeliveryGroupMethods
+        for (Integer j = (cartDeliveryGroupMethods.size() - 1); j >= 0; j--) {
+          CartExtension.CartDeliveryGroupMethod method = cartDeliveryGroupMethods.get(j);
+          cartDeliveryGroupMethods.remove(method);
+        }
+        // To clear selected Cart Delivery Group Method
+        cartDeliveryGroup.setSelectedCartDeliveryGroupMethod(null);
+  
+        // Get the Shipping Product
+        List<Product2> shippingProducts = [SELECT Id FROM Product2 WHERE ProductClass != 'VariationParent' LIMIT 1];
+        if(shippingProducts.size() == 0) {
```

Comment:
@gurpreetsainisalesforce the implementation changed here from 'all or nothing' into 'non failing partial success and partial failure at once', this will complicate further maintenance/qa/desision-taking. Should we keep it as 'all or nothing' for all the MDBs?

Reply:
This is to get the Shipping Product Id that is used to create Cart Delivery Group Methods. We re use the same shipping product for all Cart Delivery Group Methods.

Reply:
This note is general, not about a specific row. Previously we either created a new CDGM or added a CVO error, after this code change we can complete the integration with CDGMs created for some DGs and other DG having CVO errors - that is new and different. I've asked would we like to complete the calculation either with CDGMs created for all DGs or for none + CVO erros, and not mixed. WDYT? 
@gurpreetsainisalesforce @wei-liu-salesforce

Reply:
That's a good question. @wei-liu-salesforce How would to envision this?

Reply:
@gkirpichnikov After another review I think we are just adding all errors to the same cvo list and thats how it should work .. what's your thought?

Reply:
@gurpreetsainisalesforce I suggest we not create any CDGM in case of any error, so in the end we don't have semi-success/semi-failure but just a failure state (CVO error/s and no CDGM/s)

Reply:
I am not know if we want to rollback when we have an error
For example, if there are 5 deliveryGroups, and if for the first 4, deliverymethods were generated, but for the 5th there was an error, we cannot revert the deliveryMethods which we created for the last 4, we can only update the CVO
Ideally this is a sample, the customers can decide on how they act upon a CVO error, we don't need to need to have an opinion
As long as we update the CVO with the appropriate error we should fine

Reply:
Good discussion! Do we have to make a decision in this story?

Reply:
As Sneha said, this is a sample. Customers can decide what behavior they want eventually. I do see points from both angles. Especially in future if we use BuyerActionDetails in calculator, saving valid shipping result for some delivery groups can be useful.